### PR TITLE
added contacts data functions

### PR DESCRIPTION
**Pull request by ryan280**

### CODE REVIEW

**Comment by mpolimen** [comment]
Take a look at the comments I left for Brian's Application Data PR; I have a lot of the same suggestions from a quick glance of your PR

**Comment by mpolimen** [comment]
Make sure to `git pull origin main` in this branch and use the types instead of just plain old objects

File: server/data/contacts.ts [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+
+/**
+ * Get contact by id
+ * @param {string} userId
+ * @param {string} applicationId
+ * @param {string} contactId
+ * @returns {Promise<Object>} The contact given the ids above
+ */
+export const getContactById = async (userId: string, applicationId: string, contactId: string):Promise<Object> => {
```
Comment (mpolimen) [inline server/data/contacts.ts]: Object should be ContactObject

File: server/data/contacts.ts [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+    if(checkObjectId(userId)){
```
Comment (mpolimen) [inline server/data/contacts.ts]: checkObjectId returns true if the id is valid, so you want to error when it's false instead. This goes for other error checks as well...

File: server/data/contacts.ts [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    }
+    if(checkObjectId(applicationId)){
+        throw new Error('Invalid applicationId');
+    }
+    if(checkObjectId(contactId)){
+        throw new Error('Invalid contactId');
+    }
+
+    const usersCollection: any = await users();
+    const user = await usersCollection.findOne({_id: new ObjectId(userId)});
```
Comment (mpolimen) [inline server/data/contacts.ts]: UserObject

File: server/data/contacts.ts [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+    if(user === null) throw new Error('Could not find user');
+
+    const cycleId = user.cycles[user.cycles.length - 1];
```
Comment (mpolimen) [inline server/data/contacts.ts]: string

File: server/data/contacts.ts [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+    const cyclesCollection: any = await cycles();
+    const cycle = await cyclesCollection.findOne({_id: new ObjectId(cycleId)});
```
Comment (mpolimen) [inline server/data/contacts.ts]: CycleObject

File: server/data/contacts.ts [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    if(cycle === null) throw new Error('Could not find cycle');
+
+    let application = null;
```
Comment (mpolimen) [inline server/data/contacts.ts]: ApplicationObject

File: server/data/contacts.ts [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    for(let element of cycle.applications){
+        if(element._id.toString() === applicationId){
+            application = element;
+        }
+    }
+    if(application === null) throw new Error('Could not find application');
+
+    let contact = null;
```
Comment (mpolimen) [inline server/data/contacts.ts]: ContactObject

File: server/data/contacts.ts [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    return contact;
+}
+
+/**
+ * Get all contacts
+ * @param {string} userId
+ * @param {string} applicationId
+ * @returns {Promise<Object[]>}All the contacts of the application
+ */
+ export const getAllContacts = async (userId: string, applicationId: string): Promise<Object[]> => {
```
Comment (mpolimen) [inline server/data/contacts.ts]: Promise<ContactObject[]>

File: server/data/contacts.ts [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+        {
+        _id: new ObjectId(cycleId),
+        "applications._id": new ObjectId(applicationId)
+        }, 
+        {$push: {
+            "applications.$.contacts": newContact
+            }
+        }
+    );
+    if (updateInfo.WriteResult.nModified === 0) throw new Error('Could not add contact');
```
Comment (mpolimen) [inline server/data/contacts.ts]: is this the correct syntax for checking if modified? if you tested it and it works, please resolve this comment

File: server/data/index.ts [hunk 10]
```diff
@@ -1,2 +0,0 @@
-export * from './users';
```
Comment (mpolimen) [inline server/data/index.ts:2]: make sure these deleted exports come back when merging to main

**Review by mpolimen — COMMENTED** [review]
A lot of duplicate typing and error checking issues, so please go through the entire code yourself to address them all, rather than having me leave a comment for every case on here.

**Review by bwormser — APPROVED** [review]
LGTM

**Comment by roccopolimen** [comment]
You pushed a package-lock.json in the root folder of this project? Not even sure how you generated one there, but there should not be one there so delete it and make a commit with that change, and then you can merge this branch.

**Pull request merged**